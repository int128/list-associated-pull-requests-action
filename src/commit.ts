type Commit = {
  commitId: string
  committedDate: Date
}

export const findEarliestCommit = (commits: Commit[]): Commit => {
  let earliestCommit = commits[0]
  for (const commit of commits) {
    if (commit.committedDate.getTime() < earliestCommit.committedDate.getTime()) {
      earliestCommit = commit
    }
  }
  return earliestCommit
}
