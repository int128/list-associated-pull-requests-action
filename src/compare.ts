import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  base: string
  head: string
}

type Outputs = {
  commitIds: Set<string>
  earliestCommitId: string
  earliestCommitDate: Date
}

export const compareCommits = async (octokit: Octokit, inputs: Inputs): Promise<Outputs> => {
  const compareIterator = octokit.paginate.iterator(octokit.rest.repos.compareCommitsWithBasehead, {
    owner: inputs.owner,
    repo: inputs.repo,
    basehead: `${inputs.base}...${inputs.head}`,
    per_page: 100,
  })

  const commitIds = new Set<string>()
  let earliestCommitDate = new Date()
  let earliestCommitId = ''
  for await (const compare of compareIterator) {
    core.info(`Received ${compare.data.commits.length} commits of total ${compare.data.total_commits}`)
    core.info(`Remaining rate limit: ${compare.headers['x-ratelimit-remaining'] ?? '?'}`)

    for (const commit of compare.data.commits) {
      commitIds.add(commit.sha)
      if (commit.commit.committer?.date) {
        const d = new Date(commit.commit.committer.date)
        if (d < earliestCommitDate) {
          earliestCommitDate = d
          earliestCommitId = commit.sha
        }
      }
    }
  }
  return { commitIds, earliestCommitId, earliestCommitDate }
}
