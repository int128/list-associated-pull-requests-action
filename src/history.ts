import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'
import { AssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './generated/graphql'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  ref: string
  path: string
  since: Date
  sinceCommitId: string
}

export type Commit = {
  commitId: string
  pull?: {
    number: number
    author: string
  }
}

export const getCommitHistory = async (octokit: Octokit, inputs: Inputs): Promise<Commit[]> => {
  const commits: Commit[] = []
  for (let afterCursor: string | undefined; ; ) {
    const page = await getCommitHistoryByCursor(octokit, inputs, afterCursor)
    commits.push(...page.commits)
    if (page.hasNextPage === false) {
      break
    }
    afterCursor = page.endCursor
  }
  return commits
}

type CommitHistoryPage = {
  commits: Commit[]
  hasNextPage: boolean
  endCursor?: string
}

const getCommitHistoryByCursor = async (
  octokit: Octokit,
  inputs: Inputs,
  afterCursor?: string
): Promise<CommitHistoryPage> => {
  core.startGroup(
    `AssociatedPullRequestsInCommitHistoryOfSubTreeQuery (${inputs.ref}, ${inputs.path}, ${inputs.since.toISOString()})`
  )
  const q = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: inputs.owner,
    name: inputs.repo,
    expression: inputs.ref,
    path: inputs.path,
    since: inputs.since,
    after: afterCursor,
  })
  core.info(JSON.stringify(q, undefined, 2))
  core.endGroup()

  return parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(q, inputs.sinceCommitId)
}

export const parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery = (
  q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery,
  sinceCommitId: string
): CommitHistoryPage => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }

  const nodes = filterNodes(q, sinceCommitId)
  core.startGroup(`Found ${nodes.length} commit(s)`)
  const commits: Commit[] = []
  for (const node of nodes) {
    if (!node.associatedPullRequests?.nodes?.length) {
      core.info(`${node.oid} -> no pull request`)
      commits.push({ commitId: node.oid })
      continue
    }
    for (const pull of node.associatedPullRequests.nodes) {
      if (pull?.number === undefined) {
        continue
      }
      core.info(`${node.oid} -> #${pull.number}`)
      commits.push({
        commitId: node.oid,
        pull: {
          number: pull.number,
          author: pull.author?.login ?? '',
        },
      })
    }
  }
  core.endGroup()

  return {
    commits,
    hasNextPage: q.repository.object.history.pageInfo.hasNextPage,
    endCursor: q.repository.object.history.pageInfo.endCursor ?? undefined,
  }
}

const filterNodes = (q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery, sinceCommitId: string) => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }
  const nodes = []
  for (const node of q.repository.object.history.nodes ?? []) {
    if (node == null) {
      continue
    }
    nodes.push(node)
    if (node.oid === sinceCommitId) {
      break
    }
  }
  return nodes
}
