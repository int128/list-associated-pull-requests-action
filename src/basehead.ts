import * as core from '@actions/core'
import * as github from '@actions/github'
import { ChangeSet, findChangeSet } from './history'
import { getCommit } from './queries/commit'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Inputs = {
  token: string
  base: string
  head: string
  path: string
  groupBySubPaths: string[]
}

type Outputs = {
  body: string
  associatedPullRequests: number[]
}

export const computeChangeSetBetweenBaseHead = async (inputs: Inputs): Promise<Outputs> => {
  const octokit = github.getOctokit(inputs.token)

  const baseCommit = await getCommit(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.base,
  })
  core.startGroup(`Base commit of ${inputs.base}`)
  core.info(JSON.stringify(baseCommit, undefined, 2))
  core.endGroup()
  if (baseCommit.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(baseCommit.repository?.object?.__typename)} !== Commit`)
  }

  const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.head,
    path: inputs.path,
    since: baseCommit.repository.object.committedDate,
  })
  core.startGroup(`Commit history on ${inputs.head} since ${baseCommit.repository.object.committedDate}`)
  core.info(JSON.stringify(history, undefined, 2))
  core.endGroup()
  const changeSet = findChangeSet(history, baseCommit.repository.object.oid)

  if (inputs.groupBySubPaths.length === 0) {
    return {
      body: [...changeSet.pullOrCommits].map((s) => `- ${s}`).join('\n'),
      associatedPullRequests: [...changeSet.pulls],
    }
  }

  const subPathChangeSets = new Map<string, ChangeSet>()
  for (const subPath of inputs.groupBySubPaths) {
    const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
      owner: github.context.repo.owner,
      name: github.context.repo.repo,
      expression: inputs.head,
      path: subPath,
      since: baseCommit.repository.object.committedDate,
    })
    core.startGroup(`Commit history of subpath ${subPath}`)
    core.info(JSON.stringify(history, undefined, 2))
    core.endGroup()
    const changeSet = findChangeSet(history, baseCommit.repository.object.oid)
    subPathChangeSets.set(subPath, changeSet)
  }

  const otherPullOrCommits = new Set(changeSet.pullOrCommits)
  const body = []
  for (const [subPath, { pullOrCommits }] of subPathChangeSets) {
    if (pullOrCommits.size > 0) {
      body.push(`### ${subPath}`)
      for (const pullOrCommit of pullOrCommits) {
        body.push(`- ${pullOrCommit}`)
        otherPullOrCommits.delete(pullOrCommit)
      }
    }
  }
  if (otherPullOrCommits.size > 0) {
    body.push(`### Others`)
    for (const pullOrCommit of otherPullOrCommits) {
      body.push(`- ${pullOrCommit}`)
    }
  }

  return {
    body: body.join('\n'),
    associatedPullRequests: [...changeSet.pulls],
  }
}
