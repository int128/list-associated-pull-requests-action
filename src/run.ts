import * as core from '@actions/core'
import * as github from '@actions/github'
import { Association, findAssociation } from './history'
import { getCommit } from './queries/commit'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Inputs = {
  token: string
  base: string
  head: string
  path: string
  groupBySubPaths: string[]
}

type Outputs = {
  body: string
  associatedPullRequests: number[]
  pullRequestListMarkdown: string // deprecated
}

export const run = async (inputs: Inputs): Promise<Outputs> => {
  const octokit = github.getOctokit(inputs.token)

  const baseCommit = await getCommit(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.base,
  })
  core.startGroup(`Base commit of ${inputs.base}`)
  core.info(JSON.stringify(baseCommit, undefined, 2))
  core.endGroup()
  if (baseCommit.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(baseCommit.repository?.object?.__typename)} !== Commit`)
  }

  const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.head,
    path: inputs.path,
    since: baseCommit.repository.object.committedDate,
  })
  core.startGroup(`Commit history on ${inputs.head} since ${baseCommit.repository.object.committedDate}`)
  core.info(JSON.stringify(history, undefined, 2))
  core.endGroup()

  const association = findAssociation(history, baseCommit.repository.object.oid)
  if (inputs.groupBySubPaths.length === 0) {
    return {
      body: [...association.pullOrCommits].map((s) => `- ${s}`).join('\n'),
      associatedPullRequests: [...association.pulls],
      pullRequestListMarkdown: [...association.pulls].map((s) => `- #${s}`).join('\n'),
    }
  }

  const subPathAssociations = new Map<string, Association>()
  for (const subPath of inputs.groupBySubPaths) {
    const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
      owner: github.context.repo.owner,
      name: github.context.repo.repo,
      expression: inputs.head,
      path: subPath,
      since: baseCommit.repository.object.committedDate,
    })
    core.startGroup(`Commit history of subpath ${subPath}`)
    core.info(JSON.stringify(history, undefined, 2))
    core.endGroup()
    const association = findAssociation(history, baseCommit.repository.object.oid)
    subPathAssociations.set(subPath, association)
  }

  const others = new Set(association.pullOrCommits)
  const body = []
  for (const [subPath, association] of subPathAssociations) {
    body.push(`### ${subPath}`)
    for (const pullOrCommit of association.pullOrCommits) {
      body.push(`- ${pullOrCommit}`)
      others.delete(pullOrCommit)
    }
  }
  if (others.size > 0) {
    body.push(`### Others`)
    for (const pullOrCommit of others) {
      body.push(`- ${pullOrCommit}`)
    }
  }

  return {
    body: body.join('\n'),
    associatedPullRequests: [...association.pulls],
    pullRequestListMarkdown: [...association.pulls].map((s) => `- #${s}`).join('\n'),
  }
}
