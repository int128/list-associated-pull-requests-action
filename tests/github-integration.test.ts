import assert from 'assert'
import { run } from '../src/run'

const { INTEGRATION_TEST_GITHUB_TOKEN } = process.env
const describeOnlyIfToken = INTEGRATION_TEST_GITHUB_TOKEN ? describe : describe.skip

describeOnlyIfToken('GitHub integration test', () => {
  it('should generate outputs of pr-491', async () => {
    assert(INTEGRATION_TEST_GITHUB_TOKEN)

    // https://github.com/int128/list-associated-pull-requests-action/pull/491
    const outputs = await run({
      owner: 'int128',
      repo: 'list-associated-pull-requests-action',
      token: INTEGRATION_TEST_GITHUB_TOKEN,
      pullRequest: 491,
      groupByPaths: ['src', 'tests', '.github'],
      showOthersGroup: true,
    })
    expect(outputs).toMatchSnapshot()
  }, 30000)
})
