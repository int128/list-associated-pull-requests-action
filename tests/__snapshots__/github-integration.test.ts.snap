// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub integration test should generate outputs of pr-491 1`] = `
{
  "body": "### src
- #484 @int128
- #479 @renovate
- #477 @renovate
- #476 @renovate
- #474 @renovate
- #470 @renovate
- #466 @renovate
- #453 @renovate
- #463 @renovate
- #462 @renovate
- #461 @renovate
- #458 @renovate
- #456 @renovate
- #454 @renovate
- #450 @renovate
- #448 @int128
- #447 @int128
- #446 @int128
- #406 @renovate
- #444 @int128
- #437 @renovate
- #435 @renovate
- #434 @renovate
- #428 @renovate
- #427 @renovate
- #424 @renovate
- #422 @renovate
- #417 @renovate
- #413 @renovate
- #412 @renovate
- #410 @renovate
- #401 @renovate
- #382 @renovate
- #343 @renovate
- #339 @renovate
- #338 @int128
- #330 @renovate
- #206 @renovate
- #134 @renovate
- #306 @int128
- #305 @int128
- #304 @ttokutake
- #303 @int128
- #302 @int128
- #286 @int128
- #278 @int128
- #164 @int128
- #156 @int128
- #149 @renovate
- #148 @int128
- #143 @int128
- #121 @int128
- #103 @int128
- #102 @int128
- #94 @int128
- #89 @int128
- #88 @int128
- #87 @int128
- #78 @int128
- #76 @int128
- #75 @int128
- #70 @int128
- #69 @int128
- #68 @int128
- #6 @int128
### tests
- #440 @int128
- #207 @renovate
- #286 @int128
- #156 @int128
- #143 @int128
- #121 @int128
- #94 @int128
### .github
- #440 @int128
- #342 @int128
- #313 @int128
- #165 @int128
- #102 @int128
- #99 @int128
- #96 @int128
- #91 @int128
- #87 @int128
- #78 @int128
- #70 @int128
- #13 @renovate
- #9 @renovate
- #6 @int128
### Others
- #469 @renovate
- #487 @renovate
- #486 @renovate
- #485 @renovate
- #483 @renovate
- #482 @renovate
- #481 @renovate
- #480 @renovate
- #478 @renovate
- #475 @renovate
- #473 @renovate
- #472 @renovate
- #471 @renovate
- #468 @renovate
- #467 @renovate
- #465 @renovate
- #464 @renovate
- #459 @renovate
- #460 @renovate
- #457 @renovate
- #455 @renovate
- #452 @renovate
- #451 @renovate
- #449 @renovate
- #438 @renovate
- #443 @renovate
- #442 @renovate
- #441 @renovate
- #439 @renovate
- #436 @renovate
- #433 @renovate
- #432 @renovate
- #431 @renovate
- #430 @renovate
- #429 @renovate
- #426 @renovate
- #425 @renovate
- #423 @renovate
- #421 @renovate
- #420 @renovate
- #419 @renovate
- #418 @renovate
- #416 @renovate
- #415 @renovate
- #414 @renovate
- #411 @renovate
- #409 @renovate
- #408 @renovate
- #407 @renovate
- #405 @renovate
- #404 @renovate
- #403 @renovate
- #402 @renovate
- #400 @renovate
- #399 @renovate
- #398 @renovate
- #397 @renovate
- #396 @renovate
- #395 @renovate
- #394 @renovate
- #392 @renovate
- #393 @renovate
- #391 @renovate
- #390 @renovate
- #389 @renovate
- #388 @renovate
- #387 @renovate
- #386 @renovate
- #385 @renovate
- #384 @renovate
- #383 @renovate
- #381 @renovate
- #380 @renovate
- #379 @renovate
- #378 @renovate
- #377 @renovate
- #376 @renovate
- #362 @renovate
- #375 @renovate
- #374 @renovate
- #373 @renovate
- #372 @renovate
- #371 @renovate
- #370 @renovate
- #369 @renovate
- #368 @renovate
- #367 @renovate
- #366 @renovate
- #365 @renovate
- #364 @renovate
- #363 @renovate
- #361 @renovate
- #360 @renovate
- #359 @renovate
- #358 @renovate
- #357 @renovate
- #356 @renovate
- #355 @renovate
- #354 @renovate
- #353 @renovate
- #352 @renovate
- #351 @renovate
- #350 @renovate
- #349 @renovate
- #348 @renovate
- #347 @renovate
- #346 @renovate
- #345 @renovate
- #344 @renovate
- #341 @renovate
- #340 @renovate
- #333 @renovate
- #337 @renovate
- #336 @renovate
- #335 @renovate
- #334 @renovate
- #332 @renovate
- #331 @renovate
- #323 @renovate
- #329 @renovate
- #328 @renovate
- #327 @renovate
- #326 @renovate
- #325 @renovate
- #324 @renovate
- #322 @renovate
- #321 @renovate
- #320 @renovate
- #319 @renovate
- #318 @renovate
- #317 @renovate
- #316 @renovate
- #315 @renovate
- #314 @renovate
- #312 @renovate
- #311 @renovate
- #310 @renovate
- #309 @renovate
- #308 @renovate
- #307 @renovate
- #301 @renovate
- #300 @renovate
- #299 @renovate
- #298 @renovate
- #297 @renovate
- #296 @renovate
- #295 @renovate
- #294 @renovate
- #293 @renovate
- #292 @renovate
- #291 @renovate
- #290 @renovate
- #289 @renovate
- #288 @renovate
- #287 @renovate
- #285 @renovate
- #284 @renovate
- #283 @renovate
- #282 @renovate
- #281 @renovate
- #280 @renovate
- #279 @renovate
- #277 @renovate
- #276 @renovate
- #275 @renovate
- #274 @renovate
- #273 @renovate
- #272 @renovate
- #271 @renovate
- #270 @renovate
- #269 @renovate
- #268 @renovate
- #267 @renovate
- #266 @renovate
- #265 @renovate
- #264 @renovate
- #263 @renovate
- #262 @renovate
- #261 @renovate
- #260 @renovate
- #259 @renovate
- #258 @renovate
- #257 @renovate
- #256 @renovate
- #255 @renovate
- #254 @renovate
- #253 @renovate
- #252 @renovate
- #251 @renovate
- #250 @renovate
- #249 @renovate
- #248 @renovate
- #247 @renovate
- #246 @renovate
- #245 @renovate
- #244 @renovate
- #243 @renovate
- #242 @renovate
- #241 @renovate
- #240 @renovate
- #239 @renovate
- #238 @renovate
- #237 @renovate
- #236 @renovate
- #235 @renovate
- #234 @renovate
- #233 @renovate
- #232 @renovate
- #231 @renovate
- #230 @renovate
- #229 @renovate
- #228 @renovate
- #227 @renovate
- #226 @renovate
- #225 @renovate
- #224 @renovate
- #223 @renovate
- #222 @renovate
- #221 @renovate
- #220 @renovate
- #219 @renovate
- #218 @renovate
- #217 @renovate
- #216 @renovate
- #215 @renovate
- #214 @renovate
- #213 @renovate
- #212 @renovate
- #211 @renovate
- #210 @renovate
- #209 @renovate
- #208 @renovate
- #205 @renovate
- #204 @renovate
- #203 @renovate
- #202 @renovate
- #200 @renovate
- #201 @renovate
- #199 @renovate
- #198 @renovate
- #197 @renovate
- #196 @renovate
- #195 @renovate
- #194 @renovate
- #192 @renovate
- #191 @renovate
- #190 @renovate
- #189 @renovate
- #188 @renovate
- #187 @renovate
- #186 @renovate
- #185 @renovate
- #184 @renovate
- #183 @renovate
- #182 @renovate
- #181 @renovate
- #180 @renovate
- #179 @renovate
- #178 @renovate
- #177 @renovate
- #176 @renovate
- #175 @renovate
- #174 @renovate
- #173 @renovate
- #172 @renovate
- #171 @renovate
- #170 @renovate
- #169 @renovate
- #168 @renovate
- #167 @renovate
- #166 @int128
- #163 @renovate
- #162 @renovate
- #161 @renovate
- #160 @renovate
- #158 @renovate
- #159 @int128
- #157 @renovate
- #155 @renovate
- #154 @renovate
- #153 @renovate
- #152 @renovate
- #151 @renovate
- #150 @renovate
- #147 @renovate
- #146 @renovate
- #145 @renovate
- #144 @renovate
- #142 @renovate
- #141 @renovate
- #140 @renovate
- #139 @renovate
- #138 @renovate
- #137 @renovate
- #136 @renovate
- #135 @renovate
- #133 @renovate
- #132 @renovate
- #131 @renovate
- #130 @renovate
- #129 @renovate
- #128 @renovate
- #127 @renovate
- #126 @renovate
- #125 @renovate
- #124 @renovate
- #123 @renovate
- #122 @renovate
- #120 @renovate
- #119 @renovate
- #117 @renovate
- #116 @renovate
- #115 @renovate
- #114 @renovate
- #113 @renovate
- #112 @renovate
- #111 @renovate
- #106 @renovate
- #110 @int128
- #109 @renovate
- #108 @renovate
- #107 @renovate
- #105 @renovate
- #104 @renovate
- #101 @renovate
- #100 @renovate
- #97 @renovate
- #90 @renovate
- #86 @renovate
- #85 @renovate
- #84 @renovate
- #83 @renovate
- #82 @renovate
- #81 @renovate
- #80 @renovate
- #79 @renovate
- #77 @int128
- #74 @renovate
- #73 @int128
- #71 @renovate
- #67 @renovate
- #64 @renovate
- #66 @renovate
- #65 @renovate
- #63 @renovate
- #62 @renovate
- #61 @renovate
- #60 @renovate
- #59 @renovate
- #58 @renovate
- #57 @renovate
- #56 @renovate
- #55 @renovate
- #54 @renovate
- #53 @renovate
- #52 @renovate
- #51 @renovate
- #44 @renovate
- #50 @renovate
- #49 @renovate
- #48 @renovate
- #47 @renovate
- #46 @renovate
- #45 @renovate
- #43 @renovate
- #41 @renovate
- 08cc22aeb4c896c032642bbf94228bb26badb896
- #40 @renovate
- #39 @renovate
- #38 @renovate
- #37 @renovate
- #36 @renovate
- #35 @renovate
- #34 @renovate
- #33 @renovate
- #32 @renovate
- #31 @renovate
- #30 @renovate
- #29 @renovate
- #28 @renovate
- #27 @renovate
- #26 @renovate
- #25 @renovate
- #24 @renovate
- #23 @renovate
- #22 @renovate
- #21 @renovate
- #20 @renovate
- #19 @renovate
- #18 @renovate
- #17 @renovate
- #16 @renovate
- #15 @renovate
- #14 @renovate
- #12 @renovate
- #11 @renovate
- #10 @renovate
- #8 @renovate
- #7 @renovate
- #5 @renovate",
  "bodyGroups": "### src
- #484 @int128
- #479 @renovate
- #477 @renovate
- #476 @renovate
- #474 @renovate
- #470 @renovate
- #466 @renovate
- #453 @renovate
- #463 @renovate
- #462 @renovate
- #461 @renovate
- #458 @renovate
- #456 @renovate
- #454 @renovate
- #450 @renovate
- #448 @int128
- #447 @int128
- #446 @int128
- #406 @renovate
- #444 @int128
- #437 @renovate
- #435 @renovate
- #434 @renovate
- #428 @renovate
- #427 @renovate
- #424 @renovate
- #422 @renovate
- #417 @renovate
- #413 @renovate
- #412 @renovate
- #410 @renovate
- #401 @renovate
- #382 @renovate
- #343 @renovate
- #339 @renovate
- #338 @int128
- #330 @renovate
- #206 @renovate
- #134 @renovate
- #306 @int128
- #305 @int128
- #304 @ttokutake
- #303 @int128
- #302 @int128
- #286 @int128
- #278 @int128
- #164 @int128
- #156 @int128
- #149 @renovate
- #148 @int128
- #143 @int128
- #121 @int128
- #103 @int128
- #102 @int128
- #94 @int128
- #89 @int128
- #88 @int128
- #87 @int128
- #78 @int128
- #76 @int128
- #75 @int128
- #70 @int128
- #69 @int128
- #68 @int128
- #6 @int128
### tests
- #440 @int128
- #207 @renovate
- #286 @int128
- #156 @int128
- #143 @int128
- #121 @int128
- #94 @int128
### .github
- #440 @int128
- #342 @int128
- #313 @int128
- #165 @int128
- #102 @int128
- #99 @int128
- #96 @int128
- #91 @int128
- #87 @int128
- #78 @int128
- #70 @int128
- #13 @renovate
- #9 @renovate
- #6 @int128",
  "bodyOthers": "### Others
- #469 @renovate
- #487 @renovate
- #486 @renovate
- #485 @renovate
- #483 @renovate
- #482 @renovate
- #481 @renovate
- #480 @renovate
- #478 @renovate
- #475 @renovate
- #473 @renovate
- #472 @renovate
- #471 @renovate
- #468 @renovate
- #467 @renovate
- #465 @renovate
- #464 @renovate
- #459 @renovate
- #460 @renovate
- #457 @renovate
- #455 @renovate
- #452 @renovate
- #451 @renovate
- #449 @renovate
- #438 @renovate
- #443 @renovate
- #442 @renovate
- #441 @renovate
- #439 @renovate
- #436 @renovate
- #433 @renovate
- #432 @renovate
- #431 @renovate
- #430 @renovate
- #429 @renovate
- #426 @renovate
- #425 @renovate
- #423 @renovate
- #421 @renovate
- #420 @renovate
- #419 @renovate
- #418 @renovate
- #416 @renovate
- #415 @renovate
- #414 @renovate
- #411 @renovate
- #409 @renovate
- #408 @renovate
- #407 @renovate
- #405 @renovate
- #404 @renovate
- #403 @renovate
- #402 @renovate
- #400 @renovate
- #399 @renovate
- #398 @renovate
- #397 @renovate
- #396 @renovate
- #395 @renovate
- #394 @renovate
- #392 @renovate
- #393 @renovate
- #391 @renovate
- #390 @renovate
- #389 @renovate
- #388 @renovate
- #387 @renovate
- #386 @renovate
- #385 @renovate
- #384 @renovate
- #383 @renovate
- #381 @renovate
- #380 @renovate
- #379 @renovate
- #378 @renovate
- #377 @renovate
- #376 @renovate
- #362 @renovate
- #375 @renovate
- #374 @renovate
- #373 @renovate
- #372 @renovate
- #371 @renovate
- #370 @renovate
- #369 @renovate
- #368 @renovate
- #367 @renovate
- #366 @renovate
- #365 @renovate
- #364 @renovate
- #363 @renovate
- #361 @renovate
- #360 @renovate
- #359 @renovate
- #358 @renovate
- #357 @renovate
- #356 @renovate
- #355 @renovate
- #354 @renovate
- #353 @renovate
- #352 @renovate
- #351 @renovate
- #350 @renovate
- #349 @renovate
- #348 @renovate
- #347 @renovate
- #346 @renovate
- #345 @renovate
- #344 @renovate
- #341 @renovate
- #340 @renovate
- #333 @renovate
- #337 @renovate
- #336 @renovate
- #335 @renovate
- #334 @renovate
- #332 @renovate
- #331 @renovate
- #323 @renovate
- #329 @renovate
- #328 @renovate
- #327 @renovate
- #326 @renovate
- #325 @renovate
- #324 @renovate
- #322 @renovate
- #321 @renovate
- #320 @renovate
- #319 @renovate
- #318 @renovate
- #317 @renovate
- #316 @renovate
- #315 @renovate
- #314 @renovate
- #312 @renovate
- #311 @renovate
- #310 @renovate
- #309 @renovate
- #308 @renovate
- #307 @renovate
- #301 @renovate
- #300 @renovate
- #299 @renovate
- #298 @renovate
- #297 @renovate
- #296 @renovate
- #295 @renovate
- #294 @renovate
- #293 @renovate
- #292 @renovate
- #291 @renovate
- #290 @renovate
- #289 @renovate
- #288 @renovate
- #287 @renovate
- #285 @renovate
- #284 @renovate
- #283 @renovate
- #282 @renovate
- #281 @renovate
- #280 @renovate
- #279 @renovate
- #277 @renovate
- #276 @renovate
- #275 @renovate
- #274 @renovate
- #273 @renovate
- #272 @renovate
- #271 @renovate
- #270 @renovate
- #269 @renovate
- #268 @renovate
- #267 @renovate
- #266 @renovate
- #265 @renovate
- #264 @renovate
- #263 @renovate
- #262 @renovate
- #261 @renovate
- #260 @renovate
- #259 @renovate
- #258 @renovate
- #257 @renovate
- #256 @renovate
- #255 @renovate
- #254 @renovate
- #253 @renovate
- #252 @renovate
- #251 @renovate
- #250 @renovate
- #249 @renovate
- #248 @renovate
- #247 @renovate
- #246 @renovate
- #245 @renovate
- #244 @renovate
- #243 @renovate
- #242 @renovate
- #241 @renovate
- #240 @renovate
- #239 @renovate
- #238 @renovate
- #237 @renovate
- #236 @renovate
- #235 @renovate
- #234 @renovate
- #233 @renovate
- #232 @renovate
- #231 @renovate
- #230 @renovate
- #229 @renovate
- #228 @renovate
- #227 @renovate
- #226 @renovate
- #225 @renovate
- #224 @renovate
- #223 @renovate
- #222 @renovate
- #221 @renovate
- #220 @renovate
- #219 @renovate
- #218 @renovate
- #217 @renovate
- #216 @renovate
- #215 @renovate
- #214 @renovate
- #213 @renovate
- #212 @renovate
- #211 @renovate
- #210 @renovate
- #209 @renovate
- #208 @renovate
- #205 @renovate
- #204 @renovate
- #203 @renovate
- #202 @renovate
- #200 @renovate
- #201 @renovate
- #199 @renovate
- #198 @renovate
- #197 @renovate
- #196 @renovate
- #195 @renovate
- #194 @renovate
- #192 @renovate
- #191 @renovate
- #190 @renovate
- #189 @renovate
- #188 @renovate
- #187 @renovate
- #186 @renovate
- #185 @renovate
- #184 @renovate
- #183 @renovate
- #182 @renovate
- #181 @renovate
- #180 @renovate
- #179 @renovate
- #178 @renovate
- #177 @renovate
- #176 @renovate
- #175 @renovate
- #174 @renovate
- #173 @renovate
- #172 @renovate
- #171 @renovate
- #170 @renovate
- #169 @renovate
- #168 @renovate
- #167 @renovate
- #166 @int128
- #163 @renovate
- #162 @renovate
- #161 @renovate
- #160 @renovate
- #158 @renovate
- #159 @int128
- #157 @renovate
- #155 @renovate
- #154 @renovate
- #153 @renovate
- #152 @renovate
- #151 @renovate
- #150 @renovate
- #147 @renovate
- #146 @renovate
- #145 @renovate
- #144 @renovate
- #142 @renovate
- #141 @renovate
- #140 @renovate
- #139 @renovate
- #138 @renovate
- #137 @renovate
- #136 @renovate
- #135 @renovate
- #133 @renovate
- #132 @renovate
- #131 @renovate
- #130 @renovate
- #129 @renovate
- #128 @renovate
- #127 @renovate
- #126 @renovate
- #125 @renovate
- #124 @renovate
- #123 @renovate
- #122 @renovate
- #120 @renovate
- #119 @renovate
- #117 @renovate
- #116 @renovate
- #115 @renovate
- #114 @renovate
- #113 @renovate
- #112 @renovate
- #111 @renovate
- #106 @renovate
- #110 @int128
- #109 @renovate
- #108 @renovate
- #107 @renovate
- #105 @renovate
- #104 @renovate
- #101 @renovate
- #100 @renovate
- #97 @renovate
- #90 @renovate
- #86 @renovate
- #85 @renovate
- #84 @renovate
- #83 @renovate
- #82 @renovate
- #81 @renovate
- #80 @renovate
- #79 @renovate
- #77 @int128
- #74 @renovate
- #73 @int128
- #71 @renovate
- #67 @renovate
- #64 @renovate
- #66 @renovate
- #65 @renovate
- #63 @renovate
- #62 @renovate
- #61 @renovate
- #60 @renovate
- #59 @renovate
- #58 @renovate
- #57 @renovate
- #56 @renovate
- #55 @renovate
- #54 @renovate
- #53 @renovate
- #52 @renovate
- #51 @renovate
- #44 @renovate
- #50 @renovate
- #49 @renovate
- #48 @renovate
- #47 @renovate
- #46 @renovate
- #45 @renovate
- #43 @renovate
- #41 @renovate
- 08cc22aeb4c896c032642bbf94228bb26badb896
- #40 @renovate
- #39 @renovate
- #38 @renovate
- #37 @renovate
- #36 @renovate
- #35 @renovate
- #34 @renovate
- #33 @renovate
- #32 @renovate
- #31 @renovate
- #30 @renovate
- #29 @renovate
- #28 @renovate
- #27 @renovate
- #26 @renovate
- #25 @renovate
- #24 @renovate
- #23 @renovate
- #22 @renovate
- #21 @renovate
- #20 @renovate
- #19 @renovate
- #18 @renovate
- #17 @renovate
- #16 @renovate
- #15 @renovate
- #14 @renovate
- #12 @renovate
- #11 @renovate
- #10 @renovate
- #8 @renovate
- #7 @renovate
- #5 @renovate",
}
`;
